CADETProcess/optimization/surrogate.py

```diff
@@ -0,0 +1,360 @@
+from functools import wraps
+from typing import Any, NoReturn
+
+import numpy as np
+from sklearn.gaussian_process import GaussianProcessRegressor
+from sklearn.preprocessing import StandardScaler
+
+from CADETProcess.dataStructure import Structure, Typed
+from CADETProcess.optimization import Population
+
+
+class SurrogateModel(Structure):
+    """
+    Surrogate model for an evaluated population.
+
+    Attributes
+    ----------
+    population : Population
+        A population containing evaluated individuals.
+    """
+
+    population = Typed(ty=Population)
+
+    def __init__(
+            self,
+            population: Population,
+            *args, **kwargs
+            ) -> NoReturn:
+        """
+        Initialize the Surrogate Model class.
+
+        Parameters
+        ----------
+        population : Population
+            A population containing evaluated individuals.
+        """
+        super().__init__(*args, population=population, **kwargs)
+
+        self.surrogates: dict[str, dict] = {}
+        for eval_fun in [
+                'objectives',
+                'nonlinear_constraints',
+                'nonlinear_constraints_violation',
+                'meta_scores',
+                ]:
+            self.surrogates[eval_fun] = {}
+            self.surrogates[eval_fun]['gp']: GaussianProcessRegressor = None
+            self.surrogates[eval_fun]['x_scaler']: StandardScaler = None
+            self.surrogates[eval_fun]['y_scaler']: StandardScaler = None
+
+        self._update_surrogate_models()
+
+    def train_gp(
+            self,
+            X: np.ndarray,
+            Y: np.ndarray
+            ) -> tuple[GaussianProcessRegressor, StandardScaler, StandardScaler]:
+        """
+        Fit a Gaussian Process on scaled input and output.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            Feature vectors of training data (also required for prediction).
+        Y : np.ndarray
+            Target values in training data (also required for prediction).
+
+        Returns
+        -------
+        tuple[GaussianProcessRegressor, StandardScaler, StandardScaler]
+            A tuple containing the gaussian process regressor, as well as scalers for
+            input and output dimensions.
+
+        """
+        X_scaler = StandardScaler().fit(X)
+        Y_scaler = StandardScaler().fit(Y)
+
+        gpr = GaussianProcessRegressor()
```

Comment:
Consider allowing specification of hyper parameters via `kwargs` (with reasonable defaults).

See also: https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html

Important parameters:
- `kernel`: default is `ConstantKernel(1.0, constant_value_bounds="fixed") * RBF(1.0, **length_scale_bounds="fixed")` (consider moving to separate method, or expose their parameters in signature)
- `alpha`: Parameter to handle noisy data, default=1e-10